### PR TITLE
Upgrade Sphinx to 7.3.7

### LINKED ISSUE #31963: Upgrade Sphinx to version 7 to improve documentation errors
### Body

We should upgrade Sphinx to version 7

Sphinx 6 has known issues with informing about various errors in our documentation and it becomes REALLY painful. Quite often contributors are mind-boggled by the errors in their documentation after they have done apparently and innocent change. Even super-experienced engineers have hard time figuring out what the error really is and often we have to revert to some woodo-like practices like bisecting in order to narrow down and fix the problems. 

Recent examples of such problems which caused us hours of debugging are :

* https://github.com/apache/airflow/pull/30326#discussion_r1162323325 - where ` -> ' change cause hours of debugging and was only found by bisecting

* https://github.com/apache/airflow/pull/31731#discussion_r1232236952 - where lack of empty lines caused the same and also was found by bisecting

What holds us freom migration though is the lack of support for sphinx 7 in the sphinx_rtd-theme which we are using as the base for airflow theme. We build and publish our theme in https://github.com/apache/airflow-site and the sphinx_rtd_theme issue is:

* Described here https://github.com/readthedocs/sphinx_rtd_theme/issues/1463
* Proposed PR to fix it here: https://github.com/readthedocs/sphinx_rtd_theme/pull/1464

There are three ways how this can be solved:

1) make sphinx_rtd_theme release sphinx7 theme and switch to it
2) vendor in sphinx_rtd_theme into airlfow theme and make necessary changes to support sphinx7 
3) switch to another theme.
4) Possibly also switch to `sphinx.ext.autodoc` from `autoapi`

If 1) cannot be achieved, seems that 2) might be the easiest one to achieve it.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by kaxil**
closes https://github.com/apache/airflow/issues/31963

All the linked issues on the above have been resolved

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: hatch_build.py [hunk 1]
```diff
@@ -171,7 +171,7 @@
         "sphinx-jinja>=2.0.2",
         "sphinx-rtd-theme>=2.0.0",
         # Currently we are using sphinx 5 but we need to migrate to Sphinx 7
```
Comment (eladkal) [inline hatch_build.py]:
```suggestion
```

Comment (jedcunningham) [inline hatch_build.py]:
```suggestion
```

Comment (jedcunningham) [reply hatch_build.py]: Oh, I was too slow :)

**Review by jedcunningham — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Comment by kaxil** [comment]
One wrinkle:`sphinx>=7.0.1` depends on `docutils>=0.18.1,<0.21` and we have limited `docutils` to <0.17`

**Comment by kaxil** [comment]
Other related issue/PR: 
- https://github.com/apache/airflow-site/pull/843

**Comment by potiuk** [comment]
FYI @kaxil . 

From what I remember (see my old PR) the biggest problem there was to make appropriate changes to the theme so that the document "look" right. The change in docutils needed means that the generated documentation has a different HTML structure (different tags are used to construct the headers etc. ) and we need to update/rebuild CSS  and release a new theme to make it looks good (and do it in the way to not break past released documentation as the theme is shared).

**Comment by kaxil** [comment]
>From what I remember (see my old PR) the biggest problem there was to make appropriate changes to the theme so that the document "look" right. The change in docutils needed means that the generated documentation has a different HTML structure (different tags are used to construct the headers etc. ) and we need to update/rebuild CSS and release a new theme to make it looks good (and do it in the way to not break past released documentation as the theme is shared).

Yeah, I looked at the PR in https://github.com/apache/airflow-site/pull/843, planning to circle back and check with @utkarsharma2 too on the CSS changes and stuff. Not immediately but hopefully we can fix this soon'ish

**Pull request closed without merging**